docs/proposals/deployment.md

```diff
@@ -0,0 +1,119 @@
+# Deployment
+
+## Abstract
+A proposal for implementing a new resource - Deployment - which will enable
+declarative config updates for Pods and ReplicationControllers.
+
+Users will be able to create a Deployment, which will spin of
+a ReplicationController to bring up the desired pods.
+Users can also target the Deployment at existing ReplicationControllers, in
+which case the new RC will replace the existing ones.
+
+## Implementation
+
+### API Object
+The `Deployment` API object will have the following structure:
+
+```go
+type Deployment struct {
+  TypeMeta
+  ObjectMeta
+
+  // Specification of the desired behavior of the Deployment.
+  Spec DeploymentSpec
+
+  // Most recently observed status of the Deployment.
+  Status DeploymentStatus
+}
+
+type DeploymentSpec struct {
+  // Number of desired pods. This is a pointer to distinguish between explicit
+  // zero and not specified.
+  Replicas *int
+
+  // Label selector for pods. Existing ReplicationControllers which have all these
+  // label selectors will be scaled down.
+  Selector map[string][string]
+
+  // Reference to PodTemplateSpec that describes the pods that will be created.
+  TemplateRef *ObjectReference
+
+  // Describes the pods that will be created. Takes precedence over TemplateRef
+  // if specified.
+  Template *PodTemplateSpec
+
+  // The deployment strategy to use to replace existing pods with new ones.
+  Strategy DeploymentStrategy
+}
+
+type DeploymentStrategy struct {
+  // Type of deployment. Can be "Recreate" or "RollingUpdate".
+  Type DeploymentType
+
+  // Rolling update config params. Present only if DeploymentType =
+  // RollingUpdate.
+  RollingUpdate *RollingUpdateDeploymentSpec
+}
+
+type DeploymentType string
+
+const (
+  // Kill all existing pods before creating new ones.
+  DeploymentRecreate DeploymentType = "Recreate"
+
+  // Replace the old RCs by new one using rolling update i.e gradually scale down the old RCs and scale up the new one.
+  DeploymentRollingUpdate DeploymentType = "RollingUpdate"
+)
+
+// Spec to control the desired behavior of rolling update.
+type RollingUpdateDeploymentSpec struct {
+  // The percentage of pod replicas that should be updated in each interval.
+  // If nil, a fixed value of 1 pod is used.
+  UpdatePercent *int
+
+  // While updating pods, a "readiness" check is done instead of just the
+  // pod "creation" check, if this is true. False, by default.
+  UseReadinessAsAcceptance bool
+}
+
+type DeploymentStatus struct {
+  // The total number of pods created with the desired template spec.
+  Replicas int
+
+  // The most recent generation observed by the controller.
+  ObservedGeneration int64
+}
+
+```
+
+### Controller
+
+The DeploymentController will make a given Deployment happen.
+Whenever a new Deployment is created, it will:
```

Comment:
On Aug 5, 2015, at 9:21 PM, Brian Grant <notifications@github.com> wrote:

In docs/proposals/deployment.md
<https://github.com/GoogleCloudPlatform/kubernetes/pull/12236#discussion_r36373722>
:

> +}
> +
> +type DeploymentStatus struct {
> +  // The total number of pods created with the desired template spec.
> +  Replicas int
> +
> +  // The most recent generation observed by the controller.
> +  ObservedGeneration int64
> +}
> +
> +```
> +
> +### Controller
> +
> +The DeploymentController will make a given Deployment happen.
> +Whenever a new Deployment is created, it will:

Re. logs: We could get events.

Custom-code use cases:
1. Custom deployment logic
2. Post-deployment hooks


Post, pre, mid, percentage, and canary :)

—
Reply to this email directly or view it on GitHub
<https://github.com/GoogleCloudPlatform/kubernetes/pull/12236/files#r36373722>
.